app/lib/calendrical_events/provider/past_calendrical_events_page_controller.dart

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2023 Sharezone UG (haftungsbeschränkt)
+// Licensed under the EUPL-1.2-or-later.
+//
+// You may obtain a copy of the Licence at:
+// https://joinup.ec.europa.eu/software/page/eupl
+//
+// SPDX-License-Identifier: EUPL-1.2
+
+import 'dart:async';
+
+import 'package:date/date.dart';
+import 'package:flutter/material.dart';
+import 'package:group_domain_models/group_domain_models.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:sharezone/calendrical_events/analytics/past_calendrical_events_page_analytics.dart';
+import 'package:sharezone/calendrical_events/models/calendrical_event.dart';
+import 'package:sharezone/sharezone_plus/subscription_service/subscription_service.dart';
+import 'package:sharezone/timetable/src/widgets/events/event_view.dart';
+import 'package:sharezone/util/api/course_gateway.dart';
+import 'package:sharezone/util/api/school_class_gateway.dart';
+import 'package:sharezone/util/api/timetable_gateway.dart';
+
+class PastCalendricalEventsPageController extends ChangeNotifier {
+  late PastCalendricalEventsPageState state;
+  EventsSortingOrder _sortingOrder = EventsSortingOrder.descending;
+  EventsSortingOrder get sortingOrder => _sortingOrder;
+
+  StreamSubscription<List<EventView>>? _eventsSubscription;
+  late StreamSubscription<bool> _hasUnlockedSubscription;
+  final TimetableGateway timetableGateway;
+  final CourseGateway courseGateway;
+  final SchoolClassGateway schoolClassGateway;
+  final PastCalendricalEventsPageAnalytics analytics;
+
+  /// The date until which events should be fetched.
+  ///
+  /// This is used to determine which events are past events and which are
+  /// upcoming events.
+  late DateTime _getEventsUntilDateTime;
+
+  PastCalendricalEventsPageController({
+    required DateTime now,
+    required SubscriptionService subscriptionService,
+    required this.timetableGateway,
+    required this.courseGateway,
+    required this.schoolClassGateway,
+    required this.analytics,
+  }) {
+    _getEventsUntilDateTime = now;
+
+    state = PastCalendricalEventsPageLoadingState();
+    _hasUnlockedSubscription = subscriptionService
+        .hasFeatureUnlockedStream(SharezonePlusFeature.viewPastEvents)
+        .listen((hasUnlocked) {
+      if (hasUnlocked) {
+        _listenToPastEvents();
+      } else {
+        state = PastCalendricalEventsPageNotUnlockedState();
+        notifyListeners();
+      }
+    })
+      ..onError((e) {
+        _setError('$e');
+      });
+  }
```

Comment:
Since the `PastCalendricalEventsPageController` is always newly created when the page is viewed, couldn't we just access `.hasFeatureUnlocked` instead of `.hasFeatureUnlockedStream`?

Reply:
Yes, true 👍